### PR TITLE
Feature/es6 integration

### PR DESCRIPTION
**Pull request by netsensei**
Implements a searcher interface for a dedicated search engine instead of relying on postgres ts_vector.

Adds:

* A dedicated reindex command with index switching via `go run main.go reindex`
* Searching in the index via `SuggestProjects`
* Works with a facade allowing support for different engines (es6, typesense,...)

### CODE REVIEW

File: search/es6/engine.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	if err := json.NewDecoder(res.Body).Decode(responseBody); err != nil {
+		return fmt.Errorf("elastic search: error parsing the response body: %w", err)
+	}
+
+	return nil
+}
+
+func (s *engine) SuggestProjects(index, q string) (map[string]json.RawMessage, error) {
```
Comment (nics) [inline search/es6/engine.go:94]: just Search?

File: search/search.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"encoding/json"
+)
+
+type Engine interface {
+	SuggestProjects(index, q string) (map[string]json.RawMessage, error)
+}
+
+type Searcher struct {
```
Comment (nics) [inline search/search.go:11]: Why do you need both Engine and the wrapping Searcher? This can just be 1  struct.

File: search/es6/engine.go (hunk 1 above)
Comment (netsensei) [reply search/es6/engine.go:94]: This is a specific query geared towards autocomplete / suggesting projects based on a fuzzy text search.

File: repositories/repo.go [hunk 3]
```diff
@@ -99,25 +97,58 @@ func (r *Repo) DeleteProject(ctx context.Context, id string) error {
...
+		page = page + size
 	}
 
-	projects := make([]*models.Project, 0, len(rows))
-	if rows == nil {
-		return projects, nil
+	return nil
+}
+
+func (r *Repo) BetweenProjects(ctx context.Context, t1, t2 time.Time, fn func(p *models.Project) bool) error {
```
[the remark this answers is not in the document]
Comment (netsensei) [reply repositories/repo.go:140]: Taken from biblio-backoffice: `PublicationsBetween` and `DatasetsBetween` (I might want to swap that to `ProjectsBetween`)

File: cli/reindex_cmd.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+		if err != nil {
+			return err
+		}
+
+		// first indexing round
+		repo.EachProject(ctx, func(p *models.Project) bool {
+			doc := es6.NewProjectDocument(p)
+
+			d, _ := json.Marshal(doc)
+			// if err != nil {
```
[the remark this answers is not in the document]
Comment (netsensei) [reply cli/reindex_cmd.go]: done!

File: search/search.go (hunk 2 above)
Comment (nics) [reply search/search.go:11]: As long as the engines have the same interface, they can be swapped, the facade is superfluous.

File: api/v1/service.go [hunk 5]
```diff
@@ -154,16 +157,28 @@ func (s *Service) GetProject(ctx context.Context, req *GetProjectRequest) (GetPr
...
 	res := &SuggestProjectsResponse{
-		Data: make([]GetProject, 0, len(ps)),
+		Data: make([]GetProject, 0, len(hits)),
 	}
 
-	for _, p := range ps {
+	for _, id := range hits {
+		p, err := s.repo.GetProject(ctx, id)
+
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (netsensei) [reply api/v1/service.go]: No, it doesn't. Removed it.

File: cli/reindex_cmd.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+		client, err := elasticsearch.NewClient(elasticsearch.Config{
+			Addresses: strings.Split(config.Search.Conn, ","),
+		})
+
+		if err != nil {
+			return err
+		}
+
+		// init a switcher
+		settings, err := os.ReadFile("etc/search/es6/project.json")
```
[the remark this answers is not in the document]
Comment (netsensei) [reply cli/reindex_cmd.go]: The settings are only used in the `reindex` command which leans on `ugent-library/index`. The internal es6 package of this application is entirely de-coupled from the indexing logic.

It's also a serialized declaration of settings, not imperative code containing application logic. It makes sense to keep that in an `etc` directory.

I did fix the json being missing due to tuning in `.gitignore`.

File: search/search.go (hunk 2 above)
Comment (netsensei) [reply search/search.go:11]: Well [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it). I'll remove this.

Comment (netsensei) [reply search/search.go]: Done!

**Pull request merged**